pkg/logs/client/tcp/destination.go

```diff
@@ -13,113 +13,128 @@ import (
 
 	"github.com/DataDog/datadog-agent/pkg/logs/client"
 	"github.com/DataDog/datadog-agent/pkg/logs/config"
+	"github.com/DataDog/datadog-agent/pkg/logs/message"
 	"github.com/DataDog/datadog-agent/pkg/logs/metrics"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
-const (
-	warningPeriod = 1000
-)
-
 // Destination is responsible for shipping logs to a remote server over TCP.
 type Destination struct {
+	sync.Mutex
 	prefixer            *prefixer
 	delimiter           Delimiter
 	connManager         *ConnectionManager
 	destinationsContext *client.DestinationsContext
 	conn                net.Conn
 	connCreationTime    time.Time
-	inputChan           chan []byte
-	once                sync.Once
+	shouldRetry         bool
+	isRetrying          bool
 }
 
 // NewDestination returns a new destination.
-func NewDestination(endpoint config.Endpoint, useProto bool, destinationsContext *client.DestinationsContext) *Destination {
+func NewDestination(endpoint config.Endpoint, useProto bool, destinationsContext *client.DestinationsContext, shouldRetry bool) *Destination {
 	prefix := endpoint.APIKey + string(' ')
+	metrics.DestinationLogsDropped.Set(endpoint.Host, &expvar.Int{})
 	return &Destination{
 		prefixer:            newPrefixer(prefix),
 		delimiter:           NewDelimiter(useProto),
 		connManager:         NewConnectionManager(endpoint),
 		destinationsContext: destinationsContext,
+		shouldRetry:         shouldRetry,
+		isRetrying:          false,
 	}
 }
 
-// Send transforms a message into a frame and sends it to a remote server,
-// returns an error if the operation failed.
-func (d *Destination) Send(payload []byte) error {
-	if d.conn == nil {
-		var err error
-
-		// We work only if we have a started destination context
-		ctx := d.destinationsContext.Context()
-		if d.conn, err = d.connManager.NewConnection(ctx); err != nil {
-			// the connection manager is not meant to fail,
-			// this can happen only when the context is cancelled.
-			return err
+// Start reads from the input, transforms a message into a frame and sends it to a remote server,
+func (d *Destination) Start(input chan *message.Payload, output chan *message.Payload) (stopChan chan struct{}) {
+	stopChan = make(chan struct{})
+	go func() {
+		for payload := range input {
+			d.sendAndRetry(payload, output)
 		}
-		d.connCreationTime = time.Now()
-	}
-
-	metrics.BytesSent.Add(int64(len(payload)))
-	metrics.TlmBytesSent.Add(float64(len(payload)))
-	metrics.EncodedBytesSent.Add(int64(len(payload)))
-	metrics.TlmEncodedBytesSent.Add(float64(len(payload)))
+		d.Lock()
+		d.isRetrying = false
+		d.Unlock()
+		stopChan <- struct{}{}
```

Comment:
Ditto from earlier on question of possible dual stopChan invocation

Reply:
I made it directional